src/main/java/com/vmware/vipclient/i18n/messages/service/CacheService.java

```diff
@@ -101,17 +99,14 @@ public void addCacheOfStatus(Map<String, String> dataMap) {
         }
     }
 
-    public List<Locale> getCachedLocales() {
-        List<Locale> locales = new ArrayList<>();
-        Cache c = VIPCfg.getInstance().getCacheManager().getCache(VIPCfg.CACHE_L3);
-        if (c == null) {
-            return locales;
-        }
-        Set<String> cacheKeys = c.keySet();
-        for (String key: cacheKeys) {
-            locales.add(getLocaleByCachedKey(key));
+    public List<Locale> getSupportedLocalesFromCache() {
+        List<Locale> result = new LinkedList<>();
```

Comment:
Product service should consume cache service. But here cache service makes use of product service. It seems a design problem.

Reply:
You have a point. Let me move things around.

Reply:
@xiaochao8 I think it's ok to call product service here, because it's reasonable for service classes to call each other.
To avoid this, I see Jessiely need make changes to multiple files,  I think it's unnecessary. How do you think?

@jessiejuachon Not sure if it's convenient to revert the change for this?

Reply:
@Xiaochao8 , @huihuiw01 . I think you both have a point. 

Huihui, yes it should be reasonable for service classes to call each other. If you see the most recent changes. I call ProductService inside ComponentService. However, I think Xiaochao's point is more about using cache services (CacheService or FormattingCacheService) only for cache-related operations. 

I was calling ProductService.getSupportedLocales inside CacheService.  ProductService.getSupportedLocales is not just doing cache-related operation. If nothing is found in cache, it tries to retrieve the data from data store. It would be incorrect to call this method inside a CacheService method--it would be misleading as CacheService should be for cache-related operations only.

Reply:
Actually, I realized that what's needed inside CacheService.isContainComponent and CacheService.getCacheOfComponent is not the actual list of supported locales. Instead, it only needs the list of locales of currently cached messages. On other words, if a locale may supported in VIP service, but its messages are not yet in cache. Which means it should not be included in the list to pick from.